resources/smpp/ra/src/main/java/org/restcomm/slee/resource/smpp/SmppSessionsImpl.java

```diff
@@ -495,4 +492,113 @@ public void fireUnknownThrowable(Throwable throwable) {
+		}
+
+		public void deactivate() {
+			running = false;
+			this.interrupt();
+			while (!queue.isEmpty()) 
+			{
+				try {
+					SmppSendingTask task = queue.take();
+					if (task != null) 
+					{
+						Exception ex = new InterruptedException("SMPP Sending Thread was stopped");
+						fireSendPduStatusEvent(EventsType.SEND_PDU_STATUS, task.getSmppServerTransaction(), task.getRequest(), task.getResponse(), ex, false);
+					}
+				} 
+				catch (InterruptedException e) 
+				{
+				}
+			}
+		}
+
+		public void run() {
+			while (running) {
+				SmppSendingTask task = null;
+				try {
+					task = queue.take();
+					if (task != null) {
+						DefaultSmppSession defaultSmppSession = task.getEsme()
+								.getSmppSession();
+						if (task.getResponse() == null) {
+							try
+							{
+								defaultSmppSession.sendRequestPdu(task.getRequest(), task.getTimeoutMillis(), false);
+								fireSendPduStatusEvent(EventsType.SEND_PDU_STATUS, task.getSmppServerTransaction(), task.getRequest(), task.getResponse(), null, true);
+							} 
+							catch (RecoverablePduException | UnrecoverablePduException | SmppTimeoutException | SmppChannelException| InterruptedException e)
+							{
+								fireSendPduStatusEvent(EventsType.SEND_PDU_STATUS, task.getSmppServerTransaction(), task.getRequest(), task.getResponse(), e, false);
+							}
+						} else {
+							try
+							{
+								defaultSmppSession.sendResponsePdu(task.getResponse());
+								fireSendPduStatusEvent(EventsType.SEND_PDU_STATUS, task.getSmppServerTransaction(), task.getRequest(), task.getResponse(), null, true);
+							} 
+							catch (RecoverablePduException | UnrecoverablePduException | SmppChannelException | InterruptedException e) 
+							{
+								fireSendPduStatusEvent(EventsType.SEND_PDU_STATUS, task.getSmppServerTransaction(), task.getRequest(), task.getResponse(), e, false);
+							} 
+							finally 
+							{
+								SmppSessionCounters smppSessionCounters = task.getEsme().getSmppSession().getCounters();
+								SmppTransactionImpl smppTransactionImpl = (SmppTransactionImpl) task.getRequest().getReferenceObject();
+								long responseTime = System.currentTimeMillis() - smppTransactionImpl.getStartTime();
+								countSendResponsePdu(smppSessionCounters, task.getResponse(), responseTime, responseTime);
+
+								if (task.getSmppServerTransaction() == null) {
+									tracer.severe(String
+											.format("SmppTransactionImpl Activity is null while trying to send PduResponse=%s",
+													task.getResponse()));
+								} else
+									smppServerResourceAdaptor.endActivity(task
+											.getSmppServerTransaction());
+							}
+						}
+					}
+				}
+				catch (InterruptedException e) 
+				{
+					if (task != null)
+						fireSendPduStatusEvent(EventsType.SEND_PDU_STATUS, task.getSmppServerTransaction(), task.getRequest(), task.getResponse(), e, false);
+				}
+			}
+		}
+
+		public void offer(SmppSendingTask task) {
+			queue.offer(task);
```

Comment:
This operation is block able only in case there is queue size limit. Otherwise it will not get blocked.